// @beyerleinf/ngx-dnd/esm2015/lib/directives/sortable-container.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import { FormArray } from '@angular/forms';
import { DragDropConfig } from '../config/drag-drop-config';
import { DragDropSortableService } from '../service/drag-drop-sortable/drag-drop-sortable.service';
import { DragDropService } from '../service/drag-drop/drag-drop.service';
import { SortableArrayHandler, SortableFormArrayHandler } from '../util';
import { AbstractDirective } from './abstract/abstract.directive';
/* tslint:disable directive-selector */
export class SortableContainerDirective extends AbstractDirective {
    /**
     * @param {?} elementRef
     * @param {?} dragDropService
     * @param {?} config
     * @param {?} cdr
     * @param {?} sortableDataService
     */
    constructor(elementRef, dragDropService, config, cdr, sortableDataService) {
        super(elementRef, dragDropService, config, cdr);
        this.sortableDataService = sortableDataService;
        this._sortableData = [];
        this.dragEnabled = false;
    }
    /**
     * @param {?} sortableData
     * @return {?}
     */
    set sortableData(sortableData) {
        this._sortableData = sortableData;
        if (sortableData instanceof FormArray) {
            this._sortableHandler = new SortableFormArrayHandler();
        }
        else {
            this._sortableHandler = new SortableArrayHandler();
        }
        this.dropEnabled = !!this._sortableData;
    }
    /**
     * @return {?}
     */
    get sortableData() {
        return this._sortableData;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    dragEnterCallback(event) {
        if (this.sortableDataService.isDragged) {
            /** @type {?} */
            const item = this.sortableDataService.sortableContainer.getItemAt(this.sortableDataService.index);
            if (this.indexOf(item === -1)) {
                this.sortableDataService.sortableContainer.removeItemAt(this.sortableDataService.index);
                if (this.sortableDataService.sortableContainer.sortableData.length === 0) {
                    this.sortableDataService.sortableContainer.dropEnabled = true;
                }
                this.insertItemAt(item, 0);
                this.sortableDataService.sortableContainer = this;
                this.sortableDataService.index = 0;
            }
            this.detectChanges();
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    getItemAt(index) {
        return this._sortableHandler.getItemAt(this._sortableData, index);
    }
    /**
     * @param {?} item
     * @return {?}
     */
    indexOf(item) {
        return this._sortableHandler.indexOf(this._sortableData, item);
    }
    /**
     * @param {?} index
     * @return {?}
     */
    removeItemAt(index) {
        this._sortableHandler.removeItemAt(this._sortableData, index);
    }
    /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    insertItemAt(item, index) {
        this._sortableHandler.insertItemAt(this._sortableData, item, index);
    }
}
SortableContainerDirective.decorators = [
    { type: Directive, args: [{ selector: '[dnd-sortable-container]' },] }
];
/** @nocollapse */
SortableContainerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: DragDropService },
    { type: DragDropConfig },
    { type: ChangeDetectorRef },
    { type: DragDropSortableService }
];
SortableContainerDirective.propDecorators = {
    sortableData: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype._sortableData;
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype._sortableHandler;
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype.sortableDataService;
}
