// @beyerleinf/ngx-dnd/esm2015/lib/dnd.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { DragDropConfig } from './config/drag-drop-config';
import { DraggableHandleDirective } from './directives/draggable-handle.directive';
import { DraggableDirective } from './directives/draggable.directive';
import { DroppableDirective } from './directives/droppable.directive';
import { SortableContainerDirective } from './directives/sortable-container.directive';
import { SortableHandleDirective } from './directives/sortable-handle.directive';
import { SortableDirective } from './directives/sortable.directive';
export class DndModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: DndModule,
            providers: [DragDropConfig],
        };
    }
}
DndModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    DraggableDirective,
                    DraggableHandleDirective,
                    DroppableDirective,
                    SortableContainerDirective,
                    SortableDirective,
                    SortableHandleDirective,
                ],
                exports: [
                    DraggableDirective,
                    DraggableHandleDirective,
                    DroppableDirective,
                    SortableContainerDirective,
                    SortableDirective,
                    SortableHandleDirective,
                ],
            },] }
];
