// @beyerleinf/ngx-dnd/esm5/lib/directives/draggable.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Directive, ElementRef, EventEmitter, Input, Output, } from '@angular/core';
import { DragDropConfig } from '../config/drag-drop-config';
import { DragDropService } from '../service/drag-drop/drag-drop.service';
import { AbstractDirective } from './abstract/abstract.directive';
/* tslint:disable directive-selector no-output-on-prefix */
var DraggableDirective = /** @class */ (function (_super) {
    tslib_1.__extends(DraggableDirective, _super);
    function DraggableDirective(elementReference, dragDropService, config, cdr) {
        var _this = _super.call(this, elementReference, dragDropService, config, cdr) || this;
        _this.onDragStart = new EventEmitter();
        _this.onDragEnd = new EventEmitter();
        _this.onDragSuccess = new EventEmitter();
        _this.defaultCursor = _this.element.style.cursor;
        _this.dragEnabled = true;
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DraggableDirective.prototype.dragStartCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dragDropService.isDragged = true;
        this.dragDropService.dragData = this.dragData;
        this.dragDropService.onDragSuccessCallback = this.onDragSuccess;
        this.element.classList.add(this.config.onDragStartClass);
        this.onDragStart.emit({ dragData: this.dragData, mouseEvent: event });
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DraggableDirective.prototype.dragEndCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.dragDropService.isDragged = false;
        this.dragDropService.dragData = null;
        this.dragDropService.onDragSuccessCallback = null;
        this.element.classList.remove(this.config.onDragStartClass);
        this.onDragEnd.emit({ dragData: this.dragData, mouseEvent: event });
    };
    DraggableDirective.decorators = [
        { type: Directive, args: [{ selector: '[dnd-draggable]' },] }
    ];
    /** @nocollapse */
    DraggableDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DragDropService },
        { type: DragDropConfig },
        { type: ChangeDetectorRef }
    ]; };
    DraggableDirective.propDecorators = {
        dragData: [{ type: Input }],
        onDragStart: [{ type: Output }],
        onDragEnd: [{ type: Output }],
        onDragSuccess: [{ type: Output }]
    };
    return DraggableDirective;
}(AbstractDirective));
export { DraggableDirective };
if (false) {
    /** @type {?} */
    DraggableDirective.prototype.dragData;
    /** @type {?} */
    DraggableDirective.prototype.onDragStart;
    /** @type {?} */
    DraggableDirective.prototype.onDragEnd;
    /** @type {?} */
    DraggableDirective.prototype.onDragSuccess;
}
