// @beyerleinf/ngx-dnd/esm5/lib/directives/droppable.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Directive, ElementRef, EventEmitter, Output, } from '@angular/core';
import { DragDropConfig } from '../config/drag-drop-config';
import { DragDropService } from '../service/drag-drop/drag-drop.service';
import { isPresent } from '../util';
import { AbstractDirective } from './abstract/abstract.directive';
/* tslint:disable directive-selector no-output-on-prefix */
var DroppableDirective = /** @class */ (function (_super) {
    tslib_1.__extends(DroppableDirective, _super);
    function DroppableDirective(elementReference, dragDropService, config, cdr) {
        var _this = _super.call(this, elementReference, dragDropService, config, cdr) || this;
        _this.onDropSuccess = new EventEmitter();
        _this.onDragEnter = new EventEmitter();
        _this.onDragOver = new EventEmitter();
        _this.onDragLeave = new EventEmitter();
        _this.dropEnabled = true;
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DroppableDirective.prototype.dragEnterCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.dragDropService.isDragged) {
            this.element.classList.add(this.config.onDragEnterClass);
            this.onDragEnter.emit({
                dragData: this.dragDropService.dragData,
                mouseEvent: event,
            });
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DroppableDirective.prototype.dragOverCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.dragDropService.isDragged) {
            this.element.classList.add(this.config.onDragOverClass);
            this.onDragOver.emit({
                dragData: this.dragDropService.dragData,
                mouseEvent: event,
            });
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DroppableDirective.prototype.dragLeaveCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.dragDropService.isDragged) {
            this.element.classList.remove(this.config.onDragOverClass);
            this.element.classList.remove(this.config.onDragEnterClass);
            this.onDragLeave.emit({
                dragData: this.dragDropService.dragData,
                mouseEvent: event,
            });
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DroppableDirective.prototype.dropCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var dataTransfer = ((/** @type {?} */ (event))).dataTransfer;
        if (this.dragDropService.isDragged ||
            (isPresent(dataTransfer) && isPresent(dataTransfer.files))) {
            this.onDropSuccess.emit({
                dragData: this.dragDropService.dragData,
                mouseEvent: event,
            });
            if (isPresent(this.dragDropService.onDragSuccessCallback)) {
                this.dragDropService.onDragSuccessCallback.emit({
                    dragData: this.dragDropService.dragData,
                    mouseEvent: event,
                });
            }
            this.element.classList.remove(this.config.onDragOverClass);
            this.element.classList.remove(this.config.onDragEnterClass);
        }
    };
    DroppableDirective.decorators = [
        { type: Directive, args: [{ selector: '[dnd-droppable]' },] }
    ];
    /** @nocollapse */
    DroppableDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DragDropService },
        { type: DragDropConfig },
        { type: ChangeDetectorRef }
    ]; };
    DroppableDirective.propDecorators = {
        onDropSuccess: [{ type: Output }],
        onDragEnter: [{ type: Output }],
        onDragOver: [{ type: Output }],
        onDragLeave: [{ type: Output }]
    };
    return DroppableDirective;
}(AbstractDirective));
export { DroppableDirective };
if (false) {
    /** @type {?} */
    DroppableDirective.prototype.onDropSuccess;
    /** @type {?} */
    DroppableDirective.prototype.onDragEnter;
    /** @type {?} */
    DroppableDirective.prototype.onDragOver;
    /** @type {?} */
    DroppableDirective.prototype.onDragLeave;
}
