// @beyerleinf/ngx-dnd/esm5/lib/directives/sortable-container.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import { FormArray } from '@angular/forms';
import { DragDropConfig } from '../config/drag-drop-config';
import { DragDropSortableService } from '../service/drag-drop-sortable/drag-drop-sortable.service';
import { DragDropService } from '../service/drag-drop/drag-drop.service';
import { SortableArrayHandler, SortableFormArrayHandler } from '../util';
import { AbstractDirective } from './abstract/abstract.directive';
/* tslint:disable directive-selector */
var SortableContainerDirective = /** @class */ (function (_super) {
    tslib_1.__extends(SortableContainerDirective, _super);
    function SortableContainerDirective(elementRef, dragDropService, config, cdr, sortableDataService) {
        var _this = _super.call(this, elementRef, dragDropService, config, cdr) || this;
        _this.sortableDataService = sortableDataService;
        _this._sortableData = [];
        _this.dragEnabled = false;
        return _this;
    }
    Object.defineProperty(SortableContainerDirective.prototype, "sortableData", {
        get: /**
         * @return {?}
         */
        function () {
            return this._sortableData;
        },
        set: /**
         * @param {?} sortableData
         * @return {?}
         */
        function (sortableData) {
            this._sortableData = sortableData;
            if (sortableData instanceof FormArray) {
                this._sortableHandler = new SortableFormArrayHandler();
            }
            else {
                this._sortableHandler = new SortableArrayHandler();
            }
            this.dropEnabled = !!this._sortableData;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    SortableContainerDirective.prototype.dragEnterCallback = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.sortableDataService.isDragged) {
            /** @type {?} */
            var item = this.sortableDataService.sortableContainer.getItemAt(this.sortableDataService.index);
            if (this.indexOf(item === -1)) {
                this.sortableDataService.sortableContainer.removeItemAt(this.sortableDataService.index);
                if (this.sortableDataService.sortableContainer.sortableData.length === 0) {
                    this.sortableDataService.sortableContainer.dropEnabled = true;
                }
                this.insertItemAt(item, 0);
                this.sortableDataService.sortableContainer = this;
                this.sortableDataService.index = 0;
            }
            this.detectChanges();
        }
    };
    /**
     * @param {?} index
     * @return {?}
     */
    SortableContainerDirective.prototype.getItemAt = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return this._sortableHandler.getItemAt(this._sortableData, index);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    SortableContainerDirective.prototype.indexOf = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        return this._sortableHandler.indexOf(this._sortableData, item);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    SortableContainerDirective.prototype.removeItemAt = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this._sortableHandler.removeItemAt(this._sortableData, index);
    };
    /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    SortableContainerDirective.prototype.insertItemAt = /**
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    function (item, index) {
        this._sortableHandler.insertItemAt(this._sortableData, item, index);
    };
    SortableContainerDirective.decorators = [
        { type: Directive, args: [{ selector: '[dnd-sortable-container]' },] }
    ];
    /** @nocollapse */
    SortableContainerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DragDropService },
        { type: DragDropConfig },
        { type: ChangeDetectorRef },
        { type: DragDropSortableService }
    ]; };
    SortableContainerDirective.propDecorators = {
        sortableData: [{ type: Input }]
    };
    return SortableContainerDirective;
}(AbstractDirective));
export { SortableContainerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype._sortableData;
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype._sortableHandler;
    /**
     * @type {?}
     * @private
     */
    SortableContainerDirective.prototype.sortableDataService;
}
