// @beyerleinf/ngx-dnd/esm2015/lib/service/drag-drop-sortable/drag-drop-sortable.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { DragDropConfig } from '../../config/drag-drop-config';
import { isPresent } from '../../util';
import * as i0 from "@angular/core";
import * as i1 from "../../config/drag-drop-config";
export class DragDropSortableService {
    /**
     * Creates an instance of DragDropSortableService.
     * \@memberof DragDropSortableService
     * @param {?} config The DragDropConfig.
     */
    constructor(config) {
        this.config = config;
    }
    /**
     * Gets the last element that was marked sortable.
     *
     * \@readonly
     * \@memberof DragDropSortableService
     * @return {?}
     */
    get element() {
        return this._element;
    }
    /**
     * Assigns the `onSortableDragClass` to the given element.
     *
     * \@memberof DragDropSortableService
     * @param {?} e The element to assign the CSS class to.
     * @return {?}
     */
    markSortable(e) {
        if (isPresent(this._element)) {
            this._element.classList.remove(this.config.onSortableDragClass);
        }
        if (isPresent(e)) {
            this._element = e;
            this._element.classList.add(this.config.onSortableDragClass);
        }
    }
}
DragDropSortableService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
DragDropSortableService.ctorParameters = () => [
    { type: DragDropConfig }
];
/** @nocollapse */ DragDropSortableService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function DragDropSortableService_Factory() { return new DragDropSortableService(i0.ɵɵinject(i1.DragDropConfig)); }, token: DragDropSortableService, providedIn: "root" });
if (false) {
    /**
     * The last element that was marked sortable.
     *
     * \@memberof DragDropSortableService
     * @type {?}
     * @private
     */
    DragDropSortableService.prototype._element;
    /** @type {?} */
    DragDropSortableService.prototype.index;
    /** @type {?} */
    DragDropSortableService.prototype.sortableContainer;
    /** @type {?} */
    DragDropSortableService.prototype.isDragged;
    /**
     * @type {?}
     * @private
     */
    DragDropSortableService.prototype.config;
}
